lib/source/consul.js

```diff
@@ -0,0 +1,292 @@
+    this.emit('shutdown');
+  }
+
+  /**
+   * Return an object describing the Consul plugin's current status
+   *
+   * @return {Object}
+   */
+  status() {
+    return {
+      okay: this._okay,
+      running: this._running
+    };
+  }
+
+  /**
+   * Check the status of the Consul watcher
+   *
+   * @return {Boolean} True if the pooling loop is running
+   * @private
+   */
+  get _running() {
+    return !!this._consul;
+  }
+
+  /* Process updates from Consul's service/list API.
+   *
+   * The data returned Consul looks like this:
+   * [
+   *   "consul": [],
+   *   "service": ["production", "develop"]
+   * ]
+   *
+   * @param {Object} data  The JSON data from the service/list API.
+   * @private
+   */
+  _onServiceListChange(data) {
+    const services = Object.keys(data);
+    const watcherInfo = [];
+
+    services.forEach((service) => {
+      const tags = data[service];
+
+      if (tags.length > 0) {
+        tags.forEach((tag) => {
+          watcherInfo.push({
+            name: service + '-' + tag,
+            options: {
+              passing: true,
+              service,
+              tag
+            }
+          });
+        });
+      } else {
+        watcherInfo.push({
+          name: service,
+          options: {
+            passing: true,
+            service
+          }
+        });
+      }
+    });
+
+    const newWatchers = watcherInfo.filter((info) => {
+      return !this._hasHealthWatcher(info.name);
+    });
+
+    newWatchers.forEach((info) => {
+      const watcher = this._consul.watch({
+        method: this._consul.health.service,
+        options: info.options
+      });
+
+      watcher.on('error', (error) => {
+        this._error(error);
+      });
+
+      watcher.on('change', (healthServiceData) => {
+        this._onHealthServiceChange(info.name, healthServiceData);
+      });
+
+      this._registerHealthWatcher(info.name, watcher);
+    });
+
+    this._okay = true;
+  }
+
+  /**
+   * Process updates from Consul's health/service API.
+   *
+   * The data returned from Consul has the following format:
+   * [{
+   *   "Node": {
+   *     "Address": "127.0.0.1"
+   *   },
+   *   "Service": {
+   *     "Address": ""
+   *   }
+   * }]
+   *
+   * The Service.Address attribute is provided may or may not be given.
+   * If it's provided, it's the address defined by the service. The
+   * Node.Address attribute is the address provided by Consul.
+   *
+   * @param {String} name  The name of the watcher
+   * @param {Object} data  The JSON data from the health/service API
+   * @private
+   */
+  _onHealthServiceChange(name, data) {
+    if (data.length > 0) {
+      const addresses = [];
+
+      for (let i = 0; i < data.length; i += 1) {
+        const info = data[i];
+
+        if (info.Service && info.Service.Address) {
+          addresses.push(info.Service.Address);
+        } else if (info.Node && info.Node.Address) {
+          addresses.push(info.Node.Address);
+        }
```

Comment:
I need to figure out the difference between "service unregistered" and "health checks failing". If the service unregisters, we want to delete the watcher. If health checks are failing, we want to keep the watcher around.